File: android/core/data/src/main/kotlin/app/hapi/data/store/ScratchlistStore.kt
package app.hapi.data.store

import app.hapi.data.api.ApiError
import app.hapi.data.api.HapiApi
import app.hapi.protocol.wire.SCRATCHLIST_MAX_ENTRIES
import app.hapi.protocol.wire.SCRATCHLIST_MAX_TEXT_LENGTH
import app.hapi.protocol.wire.ScratchlistAttachment
import app.hapi.protocol.wire.ScratchlistAttachmentLimits
import app.hapi.protocol.wire.ScratchlistEntry
import app.hapi.protocol.wire.ScratchlistEntryCreateRequest
import app.hapi.protocol.wire.ScratchlistEntryUpdateRequest
import app.hapi.protocol.wire.ScratchlistErrorCodes
import java.util.Base64
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.cancellation.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/** Live scratchlist view of one session ([SessionScratchlist.state]). */
data class ScratchlistSessionState(
    /** Optimistic rows first, then hub order (`createdAt DESC`). */
    val entries: List<ScratchlistEntry> = emptyList(),
    /** At least one fetch succeeded — an empty [entries] means "truly empty". */
    val loaded: Boolean = false,
    val isRefreshing: Boolean = false,
    /** No successful fetch yet and the last attempt failed → error state. */
    val loadFailed: Boolean = false,
    /**
     * The session sits at the 200-entry cap (local count, or the hub's 409
     * `scratchlist_at_cap` verdict) — disable the add affordances.
     */
    val atCap: Boolean = false,
    /** Filenames of attachment uploads in flight (indeterminate progress chips). */
    val uploadsInFlight: List<String> = emptyList(),
)

sealed interface ScratchlistCreateResult {
    data class Created(val entry: ScratchlistEntry) : ScratchlistCreateResult
    data object AtCap : ScratchlistCreateResult
    data class Failed(val error: Exception) : ScratchlistCreateResult
}

sealed interface ScratchlistUploadResult {
    data class Uploaded(val attachment: ScratchlistAttachment) : ScratchlistUploadResult

    /** [code] is the typed hub code when present (`scratchlist_attachment_too_large`, …). */
    data class Failed(val message: String, val code: String? = null) : ScratchlistUploadResult
}

sealed interface ScratchlistAttachmentDeleteResult {
    data object Removed : ScratchlistAttachmentDeleteResult

    /** 409 `scratchlist_attachment_in_use` — an entry still references the file. */
    data object InUse : ScratchlistAttachmentDeleteResult
    data class Failed(val error: Exception) : ScratchlistAttachmentDeleteResult
}

/**
 * The scratchlist surface UI layers depend on ([ScratchlistStore] is the
 * production implementation; ViewModel tests substitute fakes).
 */
interface SessionScratchlist {
    fun state(sessionId: String): Flow<ScratchlistSessionState>

    fun currentState(sessionId: String): ScratchlistSessionState

    /** Marks the session observed (SSE invalidations refetch it) + refreshes. */
    fun open(sessionId: String)

    /** Unmarks observed; cached entries stay for instant re-open. */
    fun release(sessionId: String)

    /** `GET /sessions/:id/scratchlist` — replaces the cached list. Throws on failure. */
    suspend fun refresh(sessionId: String)

    /**
     * Optimistic create. The optimistic row's entryId travels in the POST, so
     * a retry after an ambiguous failure is idempotent (hub answers 200 with
     * the canonical row for a known id). Local count at the 200-entry cap
     * short-circuits to [ScratchlistCreateResult.AtCap] without a request.
     */
    suspend fun createEntry(
        sessionId: String,
        text: String,
        attachments: List<ScratchlistAttachment> = emptyList(),
    ): ScratchlistCreateResult

    /**
     * Optimistic update (null = keep; `attachments = []` clears). Rolls the
     * row back on failure; a 404 drops it and refetches (entry deleted
     * elsewhere). Returns false when the update did not stick.
     */
    suspend fun updateEntry(
        sessionId: String,
        entryId: String,
        text: String? = null,
        attachments: List<ScratchlistAttachment>? = null,
    ): Boolean

    /** Optimistic delete; restores on failure. A 404 counts as success (already gone). */
    suspend fun deleteEntry(sessionId: String, entryId: String): Boolean

    /** Base64-JSON upload; in flight it appears in [ScratchlistSessionState.uploadsInFlight]. */
    suspend fun uploadAttachment(
        sessionId: String,
        filename: String,
        bytes: ByteArray,
        mimeType: String,
    ): ScratchlistUploadResult

    suspend fun deleteAttachment(sessionId: String, attachmentId: String): ScratchlistAttachmentDeleteResult

    /** Hub attachment budgets, cached after the first success; defaults offline. */
    suspend fun limits(sessionId: String): ScratchlistAttachmentLimits
}

/**
 * Per-session scratchlist cache for one hub (tiann/hapi#893, B-M4d), the
 * native twin of `web/src/lib/use-hub-scratchlist.ts`:
 *
 * - **Fetching**: [open] (chat/scratchlist screen) refreshes; a
 *   `scratchlistUpdatedAt` SSE patch signal ([invalidations], fed by
 *   `SessionStore.scratchlistInvalidations`) refetches sessions somebody
 *   observes — the timestamp is the trigger, never data. Refetches coalesce
 *   per session like the list store's 16 ms invalidation batch.
 * - **Optimistic mutations** reconcile surgically (by entryId) instead of
 *   snapshot-restore, so a concurrent SSE refetch cannot resurrect rolled-back
 *   rows; refreshes preserve optimistic rows whose POST is still in flight.
 * - **Cap**: local pre-check at [SCRATCHLIST_MAX_ENTRIES] plus the hub's 409
 *   `scratchlist_at_cap` both surface as the friendly
 *   [ScratchlistSessionState.atCap] flag (recomputed from entry count on every
 *   refresh).
 *
 * Text is trimmed and truncated to [SCRATCHLIST_MAX_TEXT_LENGTH] (web
 * truncates rather than rejects). Entries are memory-only: the hub is the
 * durable store and every open refetches.
 */
class ScratchlistStore(
    private val api: HapiApi,
    private val scope: CoroutineScope,
    invalidations: Flow<String>? = null,
    private val refreshBatchMs: Long = REFRESH_BATCH_MS,
    private val now: () -> Long = System::currentTimeMillis,
    private val entryIdGenerator: () -> String = { "scratch-${UUID.randomUUID()}" },
) : SessionScratchlist {

    private val states = MutableStateFlow<Map<String, ScratchlistSessionState>>(emptyMap())

    /** All cached sessions; per-id observation via [state]. */
    val allStates: StateFlow<Map<String, ScratchlistSessionState>> = states.asStateFlow()

    private val observers = ConcurrentHashMap<String, Int>()
    private val refreshQueued = ConcurrentHashMap.newKeySet<String>()
    private val refreshMutexes = ConcurrentHashMap<String, Mutex>()

    /** Optimistic-create entry ids whose POST has not settled (refresh keeps them). */
    private val pendingCreates = ConcurrentHashMap.newKeySet<String>()

    @Volatile
    private var cachedLimits: ScratchlistAttachmentLimits? = null

    init {
        if (invalidations != null) {
            scope.launch {
                invalidations.collect { sessionId ->
                    if ((observers[sessionId] ?: 0) > 0) scheduleRefresh(sessionId)
                }
            }
        }
    }

    override fun state(sessionId: String): Flow<ScratchlistSessionState> =
        states.map { it[sessionId] ?: EMPTY_STATE }.distinctUntilChanged()

    override fun currentState(sessionId: String): ScratchlistSessionState =
        states.value[sessionId] ?: EMPTY_STATE

    override fun open(sessionId: String) {
        observers.merge(sessionId, 1, Int::plus)
        scheduleRefresh(sessionId)
    }

    override fun release(sessionId: String) {
        observers.computeIfPresent(sessionId) { _, count -> (count - 1).takeIf { it > 0 } }
    }

    override suspend fun refresh(sessionId: String) {
        refreshMutexes.getOrPut(sessionId) { Mutex() }.withLock {
            update(sessionId) { it.copy(isRefreshing = true) }
            val fetched = try {
                api.getScratchlist(sessionId).entries
            } catch (error: Exception) {
                update(sessionId) { st ->
                    st.copy(isRefreshing = false, loadFailed = !st.loaded)
                }
                throw error
            }
            update(sessionId) { st ->
                // Keep optimistic rows whose POST is still in flight — a
                // wholesale replace would flash them away mid-create.
                val optimistic = st.entries.filter { entry ->
                    entry.entryId in pendingCreates && fetched.none { it.entryId == entry.entryId }
                }
                withEntries(st, optimistic + fetched).copy(
                    loaded = true,
                    isRefreshing = false,
                    loadFailed = false,
                )
            }
        }
    }

    /** Coalesced fire-and-forget [refresh] (per-session 16 ms batch). */
    fun scheduleRefresh(sessionId: String) {
        if (!refreshQueued.add(sessionId)) return
        scope.launch {
            delay(refreshBatchMs)
            refreshQueued.remove(sessionId)
            try {
                refresh(sessionId)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (_: Exception) {
                // Offline burst — the next signal or screen open retries.
            }
        }
    }

    // ----------------------------------------------------------- mutations --

    override suspend fun createEntry(
        sessionId: String,
        text: String,
        attachments: List<ScratchlistAttachment>,
    ): ScratchlistCreateResult {
        val trimmed = clampText(text)
        if (trimmed.isEmpty() && attachments.isEmpty()) {
            return ScratchlistCreateResult.Failed(IllegalArgumentException("Entry requires text or attachments"))
        }
        if (currentState(sessionId).entries.size >= SCRATCHLIST_MAX_ENTRIES) {
            update(sessionId) { it.copy(atCap = true) }
            return ScratchlistCreateResult.AtCap
        }

        val stamp = now()
        val optimistic = ScratchlistEntry(
            entryId = entryIdGenerator(),
            text = trimmed,
            createdAt = stamp,
            updatedAt = stamp,
            attachments = attachments,
        )
        pendingCreates.add(optimistic.entryId)
        update(sessionId) { st -> withEntries(st, listOf(optimistic) + st.entries) }

        return try {
            val response = api.createScratchlistEntry(
                sessionId,
                ScratchlistEntryCreateRequest(
                    text = trimmed,
                    entryId = optimistic.entryId,
                    createdAt = stamp,
                    attachments = attachments.takeIf { it.isNotEmpty() },
                ),
            )
            pendingCreates.remove(optimistic.entryId)
            val canonical = response.entry
            update(sessionId) { st ->
                // Replace the optimistic row; drop any duplicate the SSE
                // refetch may have landed first (web onSuccess dedupe).
                val without = st.entries.filter {
                    it.entryId != optimistic.entryId && it.entryId != canonical.entryId
                }
                withEntries(st, listOf(canonical) + without)
            }
            ScratchlistCreateResult.Created(canonical)
        } catch (error: Exception) {
            if (error is CancellationException) throw error
            pendingCreates.remove(optimistic.entryId)
            update(sessionId) { st ->
                withEntries(st, st.entries.filter { it.entryId != optimistic.entryId })
            }
            if (error is ApiError && error.status == 409 && error.code == ScratchlistErrorCodes.AT_CAP) {
                update(sessionId) { it.copy(atCap = true) }
                // Reconcile the local count with the hub's cap verdict.
                scheduleRefresh(sessionId)
                ScratchlistCreateResult.AtCap
            } else {
                ScratchlistCreateResult.Failed(error)
            }
        }
    }

    override suspend fun updateEntry(
        sessionId: String,
        entryId: String,
        text: String?,
        attachments: List<ScratchlistAttachment>?,
    ): Boolean {
        val trimmed = text?.let(::clampText)
        if (trimmed == null && attachments == null) return false
        val previous = currentState(sessionId).entries.firstOrNull { it.entryId == entryId }
        if (previous != null) {
            val optimistic = previous.copy(
                text = trimmed ?: previous.text,
                attachments = attachments ?: previous.attachments,
                updatedAt = now(),
            )
            replaceEntry(sessionId, entryId, optimistic)
        }
        return try {
            val response = api.updateScratchlistEntry(
                sessionId,
                entryId,
                ScratchlistEntryUpdateRequest(text = trimmed, attachments = attachments),
            )
            replaceEntry(sessionId, entryId, response.entry)
            true
        } catch (error: Exception) {
            if (error is CancellationException) throw error
            if (error is ApiError && error.status == 404) {
                // Deleted elsewhere — drop the row and reconcile.
                update(sessionId) { st ->
                    withEntries(st, st.entries.filter { it.entryId != entryId })
                }
                scheduleRefresh(sessionId)
            } else if (previous != null) {
                replaceEntry(sessionId, entryId, previous)
            }
            false
        }
    }

    override suspend fun deleteEntry(sessionId: String, entryId: String): Boolean {
        val entries = currentState(sessionId).entries
        val index = entries.indexOfFirst { it.entryId == entryId }
        val removed = if (index >= 0) entries[index] else null
        if (removed != null) {
            update(sessionId) { st ->
                withEntries(st, st.entries.filter { it.entryId != entryId })
            }
        }
        return try {
            api.deleteScratchlistEntry(sessionId, entryId)
            true
        } catch (error: Exception) {
            if (error is CancellationException) throw error
            if (error is ApiError && error.status == 404) return true // already gone
            if (removed != null) {
                update(sessionId) { st ->
                    if (st.entries.any { it.entryId == entryId }) return@update st
                    val next = st.entries.toMutableList()
                    next.add(index.coerceIn(0, next.size), removed)
                    withEntries(st, next)
                }
            }
            false
        }
    }

    // --------------------------------------------------------- attachments --

    override suspend fun uploadAttachment(
        sessionId: String,
        filename: String,
        bytes: ByteArray,
        mimeType: String,
    ): ScratchlistUploadResult {
        update(sessionId) { it.copy(uploadsInFlight = it.uploadsInFlight + filename) }
        return try {
            val response = api.uploadScratchlistAttachment(
                sessionId,
                filename = filename,
                contentBase64 = Base64.getEncoder().encodeToString(bytes),
                mimeType = mimeType,
            )
            val attachment = response.attachment
            if (response.success && attachment != null) {
                ScratchlistUploadResult.Uploaded(attachment)
            } else {
                ScratchlistUploadResult.Failed(response.error ?: "Upload failed", response.code)
            }
        } catch (error: Exception) {
            if (error is CancellationException) throw error
            val api = error as? ApiError
            ScratchlistUploadResult.Failed(api?.code ?: error.message ?: "Upload failed", api?.code)
        } finally {
            update(sessionId) { it.copy(uploadsInFlight = it.uploadsInFlight - filename) }
        }
    }

    override suspend fun deleteAttachment(
        sessionId: String,
        attachmentId: String,
    ): ScratchlistAttachmentDeleteResult = try {
        api.deleteScratchlistAttachment(sessionId, attachmentId)
        ScratchlistAttachmentDeleteResult.Removed
    } catch (error: Exception) {
        when {
            error is CancellationException -> throw error
            error is ApiError && error.status == 409 &&
                error.code == ScratchlistErrorCodes.ATTACHMENT_IN_USE ->
                ScratchlistAttachmentDeleteResult.InUse
            error is ApiError && error.status == 404 -> ScratchlistAttachmentDeleteResult.Removed
            else -> ScratchlistAttachmentDeleteResult.Failed(error)
        }
    }

    override suspend fun limits(sessionId: String): ScratchlistAttachmentLimits {
        cachedLimits?.let { return it }
        return try {
            api.getScratchlistLimits(sessionId).limits.also { cachedLimits = it }
        } catch (error: Exception) {
            if (error is CancellationException) throw error
            ScratchlistAttachmentLimits.DEFAULT
        }
    }

    // ------------------------------------------------------------ internal --

    /** New entries list + recomputed [ScratchlistSessionState.atCap]. */
    private fun withEntries(
        state: ScratchlistSessionState,
        entries: List<ScratchlistEntry>,
    ): ScratchlistSessionState =
        state.copy(entries = entries, atCap = entries.size >= SCRATCHLIST_MAX_ENTRIES)

    private fun replaceEntry(sessionId: String, entryId: String, entry: ScratchlistEntry) {
        update(sessionId) { st ->
            val index = st.entries.indexOfFirst { it.entryId == entryId }
            if (index < 0) return@update st
            withEntries(st, st.entries.toMutableList().also { it[index] = entry })
        }
    }

    private fun clampText(text: String): String {
        val trimmed = text.trim()
        return if (trimmed.length > SCRATCHLIST_MAX_TEXT_LENGTH) {
            trimmed.take(SCRATCHLIST_MAX_TEXT_LENGTH)
        } else {
            trimmed
        }
    }

    /** CAS update of one session's state. */
    private fun update(
        sessionId: String,
        transform: (ScratchlistSessionState) -> ScratchlistSessionState,
    ) {
        while (true) {
            val previous = states.value
            val current = previous[sessionId] ?: EMPTY_STATE
            val next = transform(current)
            if (next === current) return
            if (states.compareAndSet(previous, previous + (sessionId to next))) return
        }
    }

    private companion object {
        val EMPTY_STATE = ScratchlistSessionState()

        /** Same batch as the list store's invalidation coalescing. */
        const val REFRESH_BATCH_MS: Long = 16
    }
}

File: ios/Packages/HapiKit/Sources/HapiClient/Stores/ScratchlistStore.swift
import Foundation
import HapiProtocol
import Observation

/// Live scratchlist view of one session (``SessionScratchlistStoring/state(_:)``).
public struct ScratchlistSessionState: Equatable, Sendable {
    /// Optimistic rows first, then hub order (`createdAt DESC`).
    public var entries: [ScratchlistEntry] = []
    /// At least one fetch succeeded — an empty ``entries`` means "truly empty".
    public var loaded = false
    public var isRefreshing = false
    /// No successful fetch yet and the last attempt failed → error state.
    public var loadFailed = false
    /// The session sits at the 200-entry cap (local count, or the hub's 409
    /// `scratchlist_at_cap` verdict) — disable the add affordances.
    public var atCap = false
    /// Filenames of attachment uploads in flight (indeterminate progress chips).
    public var uploadsInFlight: [String] = []

    public init() {}
}

public enum ScratchlistCreateResult: Sendable {
    case created(ScratchlistEntry)
    case atCap
    case failed(any Error)
}

public enum ScratchlistUploadResult: Sendable {
    case uploaded(ScratchlistAttachment)

    /// `code` is the typed hub code when present (`scratchlist_attachment_too_large`, …).
    case failed(message: String, code: String?)
}

public enum ScratchlistAttachmentDeleteResult: Sendable {
    case removed

    /// 409 `scratchlist_attachment_in_use` — an entry still references the file.
    case inUse
    case failed(any Error)
}

/// Store-side validation failures (mirror of the Android store's
/// `IllegalArgumentException` path).
public enum ScratchlistStoreError: Error, Equatable {
    /// A create with neither text nor attachments (the hub would 400 it).
    case emptyEntry
}

/// The scratchlist surface UI layers depend on (``ScratchlistStore`` is the
/// production implementation; screen-model and interactor tests substitute
/// fakes — the same interface extraction as ``SessionListStoring``).
@MainActor
public protocol SessionScratchlistStoring: AnyObject {
    /// Current per-session state; reading it inside an observation scope
    /// tracks the store's changes.
    func state(_ sessionId: String) -> ScratchlistSessionState

    /// Marks the session observed (SSE invalidations refetch it) + refreshes.
    func open(_ sessionId: String)

    /// Unmarks observed; cached entries stay for instant re-open.
    func release(_ sessionId: String)

    /// `GET /sessions/:id/scratchlist` — replaces the cached list. Throws on failure.
    func refresh(_ sessionId: String) async throws

    /// Optimistic create; see ``ScratchlistStore/createEntry(sessionId:text:attachments:)``.
    func createEntry(
        sessionId: String,
        text: String,
        attachments: [ScratchlistAttachment]
    ) async -> ScratchlistCreateResult

    /// Optimistic update (nil = keep; `attachments = []` clears). Returns
    /// false when the update did not stick.
    func updateEntry(
        sessionId: String,
        entryId: String,
        text: String?,
        attachments: [ScratchlistAttachment]?
    ) async -> Bool

    /// Optimistic delete; restores on failure. A 404 counts as success (already gone).
    func deleteEntry(sessionId: String, entryId: String) async -> Bool

    /// Base64-JSON upload; in flight it appears in ``ScratchlistSessionState/uploadsInFlight``.
    func uploadAttachment(
        sessionId: String,
        filename: String,
        data: Data,
        mimeType: String
    ) async -> ScratchlistUploadResult

    func deleteAttachment(sessionId: String, attachmentId: String) async -> ScratchlistAttachmentDeleteResult

    /// Hub attachment budgets, cached after the first success; defaults offline.
    func limits(sessionId: String) async -> ScratchlistAttachmentLimits
}

extension SessionScratchlistStoring {
    /// Text-only create convenience (protocol requirements cannot default arguments).
    public func createEntry(sessionId: String, text: String) async -> ScratchlistCreateResult {
        await createEntry(sessionId: sessionId, text: text, attachments: [])
    }
}

/// Per-session scratchlist cache for one hub (tiann/hapi#893, A-M4b) — the
/// iOS twin of the Android `ScratchlistStore` / `web/src/lib/use-hub-scratchlist.ts`:
///
/// - **Fetching**: ``open(_:)`` (chat/scratchlist screen) refreshes; a
///   `scratchlistUpdatedAt` SSE patch signal — delivered by
///   `SessionListStore.onScratchlistInvalidation` into
///   ``handleInvalidation(sessionId:)`` — refetches sessions somebody
///   observes (the timestamp is the trigger, never data). Refetches coalesce
///   per session like the list store's 16 ms invalidation batch.
/// - **Optimistic mutations** reconcile surgically (by entryId) instead of
///   snapshot-restore, so a concurrent SSE refetch cannot resurrect
///   rolled-back rows; refreshes preserve optimistic rows whose POST is still
///   in flight.
/// - **Cap**: local pre-check at ``ScratchlistCaps/maxEntries`` plus the
///   hub's 409 `scratchlist_at_cap` both surface as the friendly
///   ``ScratchlistSessionState/atCap`` flag (recomputed from entry count on
///   every entries change).
///
/// Text is trimmed and truncated to ``ScratchlistCaps/maxTextLength`` (web
/// truncates rather than rejects). Entries are memory-only: the hub is the
/// durable store and every open refetches.
@MainActor @Observable
public final class ScratchlistStore: SessionScratchlistStoring {
    /// All cached sessions, keyed by session id.
    public private(set) var states: [String: ScratchlistSessionState] = [:]

    @ObservationIgnored private let api: APIClient
    @ObservationIgnored private let refreshBatch: Duration
    @ObservationIgnored private let now: () -> Int
    @ObservationIgnored private let makeEntryId: () -> String

    @ObservationIgnored private var observers: [String: Int] = [:]
    @ObservationIgnored private var refreshQueued: Set<String> = []
    /// Per-session refresh chain — serializes overlapping refreshes so a
    /// staler response can never overwrite a fresher one (the Android store's
    /// per-session mutex).
    @ObservationIgnored private var refreshTails: [String: Task<Void, any Error>] = [:]
    /// Optimistic-create entry ids whose POST has not settled (refresh keeps them).
    @ObservationIgnored private var pendingCreates: Set<String> = []
    @ObservationIgnored private var cachedLimits: ScratchlistAttachmentLimits?

    public init(
        api: APIClient,
        refreshBatch: Duration = .milliseconds(16),
        now: @escaping () -> Int = { Int(Date().timeIntervalSince1970 * 1000) },
        makeEntryId: @escaping () -> String = { "scratch-\(UUID().uuidString)" }
    ) {
        self.api = api
        self.refreshBatch = refreshBatch
        self.now = now
        self.makeEntryId = makeEntryId
    }

    public func state(_ sessionId: String) -> ScratchlistSessionState {
        states[sessionId] ?? ScratchlistSessionState()
    }

    public func open(_ sessionId: String) {
        observers[sessionId, default: 0] += 1
        scheduleRefresh(sessionId)
    }

    public func release(_ sessionId: String) {
        guard let count = observers[sessionId] else { return }
        if count <= 1 {
            observers.removeValue(forKey: sessionId)
        } else {
            observers[sessionId] = count - 1
        }
    }

    /// `scratchlistUpdatedAt` arrived in the session's SSE patch: refetch if
    /// anybody observes this session (wired by `HubSession` from
    /// `SessionListStore.onScratchlistInvalidation`).
    public func handleInvalidation(sessionId: String) {
        guard observers[sessionId, default: 0] > 0 else { return }
        scheduleRefresh(sessionId)
    }

    // MARK: - Fetching

    public func refresh(_ sessionId: String) async throws {
        let previous = refreshTails[sessionId]
        let task = Task<Void, any Error> { [weak self] in
            if let previous {
                _ = try? await previous.value
            }
            guard let self else { return }
            try await self.performRefresh(sessionId)
        }
        refreshTails[sessionId] = task
        defer {
            if refreshTails[sessionId] == task {
                refreshTails[sessionId] = nil
            }
        }
        try await task.value
    }

    /// Coalesced fire-and-forget ``refresh(_:)`` (per-session 16 ms batch,
    /// like `SessionListStore.scheduleRefresh`).
    public func scheduleRefresh(_ sessionId: String) {
        guard !refreshQueued.contains(sessionId) else { return }
        refreshQueued.insert(sessionId)
        Task { [refreshBatch] in
            try? await Task.sleep(for: refreshBatch)
            self.refreshQueued.remove(sessionId)
            // Offline burst — the next signal or screen open retries.
            try? await self.refresh(sessionId)
        }
    }

    private func performRefresh(_ sessionId: String) async throws {
        update(sessionId) { $0.isRefreshing = true }
        let fetched: [ScratchlistEntry]
        do {
            fetched = try await api.scratchlistEntries(sessionId: sessionId).entries
        } catch {
            update(sessionId) { state in
                state.isRefreshing = false
                state.loadFailed = !state.loaded
            }
            throw error
        }
        update(sessionId) { state in
            // Keep optimistic rows whose POST is still in flight — a
            // wholesale replace would flash them away mid-create.
            let optimistic = state.entries.filter { entry in
                pendingCreates.contains(entry.entryId)
                    && !fetched.contains { $0.entryId == entry.entryId }
            }
            state.entries = optimistic + fetched
            state.atCap = state.entries.count >= ScratchlistCaps.maxEntries
            state.loaded = true
            state.isRefreshing = false
            state.loadFailed = false
        }
    }

    // MARK: - Mutations

    /// Optimistic create. The optimistic row's entryId travels in the POST,
    /// so a retry after an ambiguous failure is idempotent (hub answers 200
    /// with the canonical row for a known id). Local count at the 200-entry
    /// cap short-circuits to ``ScratchlistCreateResult/atCap`` without a
    /// request.
    public func createEntry(
        sessionId: String,
        text: String,
        attachments: [ScratchlistAttachment]
    ) async -> ScratchlistCreateResult {
        let trimmed = Self.clampText(text)
        if trimmed.isEmpty && attachments.isEmpty {
            return .failed(ScratchlistStoreError.emptyEntry)
        }
        if state(sessionId).entries.count >= ScratchlistCaps.maxEntries {
            update(sessionId) { $0.atCap = true }
            return .atCap
        }

        let stamp = now()
        let optimistic = ScratchlistEntry(
            entryId: makeEntryId(),
            text: trimmed,
            createdAt: stamp,
            updatedAt: stamp,
            attachments: attachments
        )
        pendingCreates.insert(optimistic.entryId)
        setEntries(sessionId, [optimistic] + state(sessionId).entries)

        do {
            let response = try await api.createScratchlistEntry(
                sessionId: sessionId,
                ScratchlistEntryCreateRequest(
                    text: trimmed,
                    entryId: optimistic.entryId,
                    createdAt: stamp,
                    attachments: attachments.isEmpty ? nil : attachments
                )
            )
            pendingCreates.remove(optimistic.entryId)
            let canonical = response.entry
            // Replace the optimistic row; drop any duplicate the SSE refetch
            // may have landed first (web onSuccess dedupe).
            let without = state(sessionId).entries.filter {
                $0.entryId != optimistic.entryId && $0.entryId != canonical.entryId
            }
            setEntries(sessionId, [canonical] + without)
            return .created(canonical)
        } catch {
            // Roll back even when the surrounding task was cancelled — the
            // idempotent entryId means a refetch reconciles a send that
            // actually landed (the Android store rethrows cancellation
            // instead; this store's callers never rely on that).
            pendingCreates.remove(optimistic.entryId)
            setEntries(sessionId, state(sessionId).entries.filter { $0.entryId != optimistic.entryId })
            if let apiError = error as? APIError,
               apiError.status == 409, apiError.code == ScratchlistErrorCode.atCap {
                update(sessionId) { $0.atCap = true }
                // Reconcile the local count with the hub's cap verdict.
                scheduleRefresh(sessionId)
                return .atCap
            }
            return .failed(error)
        }
    }

    /// Optimistic update (nil = keep; `attachments = []` clears). Rolls the
    /// row back on failure; a 404 drops it and refetches (entry deleted
    /// elsewhere). Returns false when the update did not stick.
    public func updateEntry(
        sessionId: String,
        entryId: String,
        text: String?,
        attachments: [ScratchlistAttachment]?
    ) async -> Bool {
        let trimmed = text.map(Self.clampText)
        if trimmed == nil && attachments == nil { return false }
        let previous = state(sessionId).entries.first { $0.entryId == entryId }
        if var optimistic = previous {
            if let trimmed { optimistic.text = trimmed }
            if let attachments { optimistic.attachments = attachments }
            optimistic.updatedAt = now()
            replaceEntry(sessionId, entryId: entryId, with: optimistic)
        }
        do {
            let response = try await api.updateScratchlistEntry(
                sessionId: sessionId,
                entryId: entryId,
                ScratchlistEntryUpdateRequest(text: trimmed, attachments: attachments)
            )
            replaceEntry(sessionId, entryId: entryId, with: response.entry)
            return true
        } catch {
            if let apiError = error as? APIError, apiError.status == 404 {
                // Deleted elsewhere — drop the row and reconcile.
                setEntries(sessionId, state(sessionId).entries.filter { $0.entryId != entryId })
                scheduleRefresh(sessionId)
            } else if let previous {
                replaceEntry(sessionId, entryId: entryId, with: previous)
            }
            return false
        }
    }

    public func deleteEntry(sessionId: String, entryId: String) async -> Bool {
        let entries = state(sessionId).entries
        let index = entries.firstIndex { $0.entryId == entryId }
        let removed = index.map { entries[$0] }
        if removed != nil {
            setEntries(sessionId, entries.filter { $0.entryId != entryId })
        }
        do {
            try await api.deleteScratchlistEntry(sessionId: sessionId, entryId: entryId)
            return true
        } catch {
            if let apiError = error as? APIError, apiError.status == 404 {
                return true // already gone
            }
            if let removed, let index {
                update(sessionId) { state in
                    guard !state.entries.contains(where: { $0.entryId == entryId }) else { return }
                    state.entries.insert(removed, at: min(max(index, 0), state.entries.count))
                    state.atCap = state.entries.count >= ScratchlistCaps.maxEntries
                }
            }
            return false
        }
    }

    // MARK: - Attachments

    public func uploadAttachment(
        sessionId: String,
        filename: String,
        data: Data,
        mimeType: String
    ) async -> ScratchlistUploadResult {
        update(sessionId) { $0.uploadsInFlight.append(filename) }
        defer {
            update(sessionId) { state in
                if let index = state.uploadsInFlight.firstIndex(of: filename) {
                    state.uploadsInFlight.remove(at: index)
                }
            }
        }
        do {
            let response = try await api.uploadScratchlistAttachment(
                sessionId: sessionId,
                filename: filename,
                data: data,
                mimeType: mimeType
            )
            if response.success, let attachment = response.attachment {
                return .uploaded(attachment)
            }
            return .failed(message: response.error ?? "Upload failed", code: response.code)
        } catch {
            let apiError = error as? APIError
            let message = apiError?.code
                ?? (error as? LocalizedError)?.errorDescription
                ?? "Upload failed"
            return .failed(message: message, code: apiError?.code)
        }
    }

    public func deleteAttachment(
        sessionId: String,
        attachmentId: String
    ) async -> ScratchlistAttachmentDeleteResult {
        do {
            try await api.deleteScratchlistAttachment(sessionId: sessionId, attachmentId: attachmentId)
            return .removed
        } catch {
            if let apiError = error as? APIError {
                if apiError.status == 409, apiError.code == ScratchlistErrorCode.attachmentInUse {
                    return .inUse
                }
                if apiError.status == 404 {
                    return .removed
                }
            }
            return .failed(error)
        }
    }

    public func limits(sessionId: String) async -> ScratchlistAttachmentLimits {
        if let cachedLimits { return cachedLimits }
        do {
            let fetched = try await api.scratchlistLimits(sessionId: sessionId).limits
            cachedLimits = fetched
            return fetched
        } catch {
            return .defaultLimits
        }
    }

    // MARK: - Internal

    /// New entries list + recomputed ``ScratchlistSessionState/atCap``.
    private func setEntries(_ sessionId: String, _ entries: [ScratchlistEntry]) {
        update(sessionId) { state in
            state.entries = entries
            state.atCap = entries.count >= ScratchlistCaps.maxEntries
        }
    }

    private func replaceEntry(_ sessionId: String, entryId: String, with entry: ScratchlistEntry) {
        update(sessionId) { state in
            guard let index = state.entries.firstIndex(where: { $0.entryId == entryId }) else { return }
            state.entries[index] = entry
        }
    }

    private func update(
        _ sessionId: String,
        _ transform: (inout ScratchlistSessionState) -> Void
    ) {
        var state = states[sessionId] ?? ScratchlistSessionState()
        transform(&state)
        states[sessionId] = state
    }

    /// Trim, then truncate to the hub's UTF-16 length cap without splitting a
    /// grapheme (the web slices JS strings, which count UTF-16 units). Pure —
    /// `nonisolated` so it can be passed as an `Optional.map` transform.
    nonisolated static func clampText(_ text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.utf16.count > ScratchlistCaps.maxTextLength else { return trimmed }
        var used = 0
        var end = trimmed.startIndex
        for index in trimmed.indices {
            let next = trimmed.index(after: index)
            used += trimmed[index...index].utf16.count
            if used > ScratchlistCaps.maxTextLength { break }
            end = next
        }
        return String(trimmed[..<end])
    }
}
